csrc/scheduler/normalization_utils.cpp

```diff
@@ -721,5 +728,620 @@ getOptionalInnerOuterPersistentBufferBatches(
   }
 }
 
+int64_t getAvailableSmemSize(
+    SchedulerRuntimeInfo& runtime_info,
+    const std::vector<TensorView*>& persistent_buffers) {
+  const auto dev_prop = at::cuda::getCurrentDeviceProperties();
+  const int64_t max_shared_memory_size =
+      (int64_t)dev_prop->sharedMemPerBlockOptin;
+  // Some shared memories are reserved for kernel launch overhead and
+  // reduction_broadcast_workspace. Estimation is conservative, but should
+  // be good enough. The actual threads per block is set in the heuristics
+  // and it may be smaller than maxThreadsPerBlock.
+  // TODO: More accurate estimation of available shared memory size.
+  const int64_t kernel_overhead = (int64_t)dev_prop->reservedSharedMemPerBlock;
+  int64_t max_buffer_dtype_size = 1;
+  for (auto tv : persistent_buffers) {
+    max_buffer_dtype_size = std::max(
+        max_buffer_dtype_size,
+        dataTypeSize(tv->getDataType().value(), runtime_info.getIndexType()));
+  }
+  const int64_t reduction_broadcast_workspace =
+      (int64_t)(dev_prop->maxThreadsPerBlock) * max_buffer_dtype_size;
+  const int64_t available_shared_memory_size =
+      max_shared_memory_size - kernel_overhead - reduction_broadcast_workspace;
+  return available_shared_memory_size;
+}
+
+int64_t getPersistentBufferSize(
+    Fusion* fusion,
+    SchedulerRuntimeInfo& runtime_info,
+    HeuristicSummary* data_cache,
+    const scheduler_utils::PersistentBufferInfo& persistent_buffer_info) {
+  auto persistent_buffer_size_info = scheduler_utils::persistentBufferSize(
+      fusion, runtime_info, persistent_buffer_info, data_cache);
+
+  // Note that projected buffer size can be zero
+  auto persistent_buffer_size =
+      persistent_buffer_size_info.projected_persistent_buffer_size == 0
+      ? persistent_buffer_size_info.persistent_buffer_size
+      : std::min(
+            persistent_buffer_size_info.persistent_buffer_size,
+            persistent_buffer_size_info.projected_persistent_buffer_size);
+
+  return persistent_buffer_size;
+}
+
+// Get the appropriate scheduler based on reduction type
+std::optional<ScheduleHeuristic> getOptionalPersistentScheduleHeuristic(
```

Comment:
check if the given fusion can be scheduled by a persistent heruristics. Currently used in `TranslateApplicableWelford::isValidPersistentFusion`.

Reply:
This function can be further used in `SchedulerEntry::proposeHeuristics`. Instead of looping over all the heuristics, we can directly go to inner, outer, innerOuter, or NonePersistent based on the returned val of this function.